src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -176,10 +183,37 @@ def prune_heads(self, heads):
         self.pruned_heads = self.pruned_heads.union(heads)
 
     def _attn(self, query, key, value, attention_mask=None, head_mask=None):
-        attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
-        if self.scale_attn_weights:
-            attn_weights = attn_weights / (float(value.size(-1)) ** 0.5)
+        if self.reorder_and_upcast_attn:
```

Comment:
In general I think I'm fine with this addition as it seems to be backwards compatible. However, it would be cleaner IMO if we could define a new function `def _upcast_and_reordered_attn(elf, query, key, value, attention_mask=None, head_mask=None)` that includes all the required changes and then add a:

```python
if self.reorder_and_upcast_attn:
    attn_output, attn_weights = self._upcast_and_reordered_attn(query, key, value, attention_mask, head_mask)
else:
    attn_output, attn_weights = self._attn(query, key, value, attention_mask, head_mask)
```

=> this would be cleaner IMO because:

- Disentangle a bit the old code and new code so that readers can be assured easily that the new code with default setting is 1-to-1 the same as before
- Remove one if-else layer from this function

Reply:
Disentangled, in its own function now!